### ISSUE TITLE
powerline fonts don't render at all

### ISSUE THREAD

**Issue by ExpandingMan** [alacritty/alacritty#2657]
I know "not working" issues without much further detail are pretty annoying, but I'm pretty stumped here.  Haven't been able to get powerline working at all.  I'm on i3 on X11, I have exactly the same issue on both Manjaro and Ubuntu.  The powerline glyphs just do not render at all anywhere.  I only have this problem on alacritty, they seem to work fine on everything else, including other terminals.  I've installed every powerline package I can find in the OS repositories.

**Comment by kchibisov** [comment]
How did you set your font in `alacritty.yml`? Does it matches the output from `fc-list`? Do you have something in `.config/fontconfig/fonts.conf`? Could you share some test case: powerline symbols  + font (just the name)?

**Comment by ExpandingMan** [comment]
I'm very unsure about what this should look like in the `alacritty.yml`, I probably don't have it set up correctly.  Right now I just have it set to using `Monospace` and am just sort of hoping that it knows that powerline should be a fall-back (which might sound crazy, but it has precedent since other terminals seem to know this).

Right now I just have
```yaml
font:
  normal:
    family: Monospace
    style: Regular
```

I'm not seeing anything about `Monospace` or `Powerline` in my `fc-list`, so perhaps that's an indication that something is wrong?  The `Monospace` font seems to be working fine.

I don't have anything in `.config/fontconfig`.

**Comment by chrisduerr** [comment]
Which of your fonts is providing the powerline symbols? Also, when you say they don't render at all, you mean that there's not even a missing symbol glyph, right?

**Comment by ExpandingMan** [comment]
To be honest I'm not even sure how to check exactly which font the powerline symbols are coming from in the other terminals.  As you can tell, I'm very ignorant about how font setups typically work (I know that there is a fallback for character codes that are not found in the primary font, but I don't know how that's configured).  If you can tell me how to determine this I can tell you.

And yes, there is not even a missing symbol glyph.

**Comment by kchibisov** [comment]
> To be honest I'm not even sure how to check exactly which font the powerline symbols are coming from in the other terminals. As you can tell, I'm very ignorant about how font setups typically work (I know that there is a fallback for character codes that are not found in the primary font, but I don't know how that's configured). 

@ExpandingMan Ok, did you installed these fonts from system package manager?

**Comment by ExpandingMan** [comment]
Yes, I installed *all* of my fonts through the system package managers.  Again, same behaviors on both Manjaro and PopOS.

**Comment by kchibisov** [comment]
@ExpandingMan so, run `fc-list > current.txt`,  then remove the packages that you've installed to get powerline fonts  and run the `fc-list > no-power.txt` , after that you can `diff current.txt no-power.txt`. You can install them again after collecting the diff. Post the diff output here.

**Comment by ExpandingMan** [comment]
Ok, on my Pop_OS machine this time... (sorry for the inconsistency).  I just realized that on Pop_OS there are glyphs, but they are simply wrong, where as in Manjaro no glyphs render at all.

On Pop_OS I had the `fonts-powerline` package installed, but I don't actually know what it's doing because the fonts seem to work with that package removed (at least if I opened new terminals it worked, perhaps I would have had to reboot to clear them from memory?).

That package adds the following to `fc-list`
```
> /usr/share/fonts/opentype/PowerlineSymbols.otf: PowerlineSymbols:style=Medium
```

Pretty sure font was never in my `fc-list` on Manjaro at all, I will check it when I get a chance.

**Comment by kchibisov** [comment]
> at least if I opened new terminals it worked, perhaps I would have had to reboot to clear them from memory?

No, when you've removed that font, your font fallback changed, so you've picked up these glyphs from a different font.

> That package adds the following to `fc-list`

Could you try adding this to your `alacritty.yml`? This will probably break things, but take a look only at powerline glyphs. In this case we can understand that alacritty can actually render them, in case you'll find it hard to focus on glyphs, let me know, I'll provide more configuration for you.

```
font:
  normal:
    family: PowerlineSymbols
```
P.s.
It doesn't matter Pop_OS or Manjaro, the font picking mechanism is done by `fontconfig`. On manjaro you probably have some font that alacritty can't render atm, like `Noto Color Emoji`, since this font only has colored bitmaps and there is no way to display it in monochromatic ones unlike twitter color emoji.

P.P.s.
`fc-list` just lists the fonts on your system from some system paths (e.g `/usr/share/fonts`) , the bold part is a `family` option for `alacritty`
/usr/share/fonts/opentype/PowerlineSymbols.otf: **PowerlineSymbols**:style=Medium

P.P.P.s
In case you want a font with a lot of glyphs, you can take a look at [nerd fonts](https://github.com/ryanoasis/nerd-fonts) (powerline symbols are a subset in them). They are know to work in alacritty ( I'm using one of them) and you'll eliminate all font fallback thing.

**Comment by AkechiShiro** [comment]
I am currently having kinda of the same issue on PopOS, here is a screenshot, of what I can see on Alacritty and on my terminal and also on a Alacritty inside a tmux.
I have installed the powerlines fonts package through the system package manager.
![AlacrittyTerminal](https://user-images.githubusercontent.com/14914796/70634415-7aa81600-1c32-11ea-8314-f31bebb22c15.png)
![AlacrittywithTmux](https://user-images.githubusercontent.com/14914796/70634417-7c71d980-1c32-11ea-954d-e3dcf2dd59a6.png)

**Comment by chrisduerr** [comment]
Does it work when launching Alacritty from a terminal that displays these correctly?

**Comment by AkechiShiro** [comment]
Nope still the same issue, I can provide more logs if needed, I ran `alacritty -vvv` from a terminal that displays correctly.
And there was this line that caught my attention but nothing weird about fonts.

```
[2019-12-11 20:34] [INFO] Welcome to Alacritty                                                                                                                
[2019-12-11 20:34] [DEBUG] Estimated DPR: 1.1979166666666667
[2019-12-11 20:34] [DEBUG] Estimated Cell Size: 11 x 20
[2019-12-11 20:34] [DEBUG] Estimated Dimensions: None
[2019-12-11 20:34] [INFO] Device pixel ratio: 1.1979166666666667
[2019-12-11 20:34] [INFO] Initializing glyph cache...
[2019-12-11 20:34] [INFO] ... finished initializing glyph cache in 0.016119767s
[2019-12-11 20:34] [INFO] Cell Size: 11 x 20
[2019-12-11 20:34] [INFO] Padding: 0 x 0
[2019-12-11 20:34] [INFO] Width: 958, Height: 719
[2019-12-11 20:34] [INFO] PTY Dimensions: Line(35) x Column(87)
```

**Comment by chrisduerr** [comment]
@AkechiShiro What terminfo do you use both inside and outside of tmux?

Also @nixpulvis does this picture (with Z's + dot above and stuff) look familiar to you? I believe I've seen this in the past, but don't remember if we solved it.

**Comment by AkechiShiro** [comment]
@chrisduerr I believed, I used the one generated inside [/usr/share/terminfo/a/alacritty](https://gist.github.com/AkechiShiro/ed73634292c9acb1a9a8959951214e2e) however I may be wrong, I'm probably using the one given by `infocmp` and got this in Alacritty with Tmux : 

```
#       Reconstructed via infocmp from file: /lib/terminfo/s/screen-256color
screen-256color|GNU Screen with 256 colors,
        am, km, mir, msgr, xenl,
        colors#0x100, cols#80, it#8, lines#24, pairs#0x10000,
        acsc=++\,\,--..00``aaffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~,
        bel=^G, blink=\E[5m, bold=\E[1m, cbt=\E[Z, civis=\E[?25l,
        clear=\E[H\E[J, cnorm=\E[34h\E[?25h, cr=\r,
        csr=\E[%i%p1%d;%p2%dr, cub=\E[%p1%dD, cub1=^H,
        cud=\E[%p1%dB, cud1=\n, cuf=\E[%p1%dC, cuf1=\E[C,
        cup=\E[%i%p1%d;%p2%dH, cuu=\E[%p1%dA, cuu1=\EM,
        cvvis=\E[34l, dch=\E[%p1%dP, dch1=\E[P, dim=\E[2m,
        dl=\E[%p1%dM, dl1=\E[M, ed=\E[J, el=\E[K, el1=\E[1K,
        enacs=\E(B\E)0, flash=\Eg, home=\E[H, hpa=\E[%i%p1%dG,
        ht=^I, hts=\EH, ich=\E[%p1%d@, il=\E[%p1%dL, il1=\E[L,
        ind=\n, indn=\E[%p1%dS, is2=\E)0, kbs=^?, kcbt=\E[Z,
        kcub1=\EOD, kcud1=\EOB, kcuf1=\EOC, kcuu1=\EOA,
        kdch1=\E[3~, kend=\E[4~, kf1=\EOP, kf10=\E[21~,
        kf11=\E[23~, kf12=\E[24~, kf2=\EOQ, kf3=\EOR, kf4=\EOS,
        kf5=\E[15~, kf6=\E[17~, kf7=\E[18~, kf8=\E[19~, kf9=\E[20~,
        khome=\E[1~, kich1=\E[2~, kmous=\E[M, knp=\E[6~, kpp=\E[5~,
        nel=\EE, op=\E[39;49m, rc=\E8, rev=\E[7m, ri=\EM, rin@,
        rmacs=^O, rmcup=\E[?1049l, rmir=\E[4l, rmkx=\E[?1l\E>,
        rmso=\E[23m, rmul=\E[24m, rs2=\Ec\E[?1000l\E[?25h,
        sc=\E7,
        setab=\E[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m,
        setaf=\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m,
        sgr=\E[0%?%p6%t;1%;%?%p1%t;3%;%?%p2%t;4%;%?%p3%t;7%;%?%p4%t;5%;%?%p5%t;2%;m%?%p9%t\016%e\017%;,
        sgr0=\E[m\017, smacs=^N, smcup=\E[?1049h, smir=\E[4h,
        smkx=\E[?1h\E=, smso=\E[3m, smul=\E[4m, tbc=\E[3g,
        vpa=\E[%i%p1%dd,

```

But when I actually type `infocmp` inside alacritty with no tmux at all, I get this [output](https://gist.github.com/AkechiShiro/e42dc6294dd8415ca8f992a5cd20907d) which is quite different.

It seems that inside of a tmux I use `screen-256color` and outside of a tmux, I use `xterm-256-color`.

**Comment by chrisduerr** [comment]
@AkechiShiro Do the powerline symbols show up outside of tmux?

**Comment by AkechiShiro** [comment]
@chrisduerr They don't, they show up the same way inside Alacritty with or without tmux, it doesn't seem to be an issue related with using tmux or not, it looks like it's somewhere else.

**Comment by bric3** [comment]
I have the same issue using the `Sauce Code Pro Nerd Font Complete Mono`

```
  ~ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.15.2
BuildVersion:	19C57
```


Alacritty-v0.4.0


<details>
<summary>`/Applications/Alacritty.app/Contents/MacOS/alacritty -vvv` logs</summary>
```
Created log file at "/var/folders/hp/bqc3d0yj2w7ds87z242lwn100000gp/T/Alacritty-29233.log"
[2019-12-16 14:45] [INFO] Welcome to Alacritty
[2019-12-16 14:45] [INFO] Configuration loaded from "/Users/bric3/.config/alacritty/alacritty.yml"
[2019-12-16 14:45] [DEBUG] Estimated DPR: 2
[2019-12-16 14:45] [DEBUG] Estimated Cell Size: 13 x 26
[2019-12-16 14:45] [DEBUG] Estimated Dimensions: None
[2019-12-16 14:45] [INFO] Device pixel ratio: 2
[2019-12-16 14:45] [INFO] Initializing glyph cache...
[2019-12-16 14:45] [INFO] ... finished initializing glyph cache in 0.035702094s
[2019-12-16 14:45] [INFO] Cell Size: 13 x 26
[2019-12-16 14:45] [INFO] Padding: 0 x 0
[2019-12-16 14:45] [INFO] Width: 1600, Height: 1200
[2019-12-16 14:45] [INFO] PTY Dimensions: Line(46) x Column(123)
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/color.rs:35] Scaling RGB by 0.66 from Rgb { r: 234, g: 234, b: 234 } to Rgb { r: 154, g: 154, b: 154 }
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/color.rs:204] Deriving dim colors from normal colors
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/color.rs:35] Scaling RGB by 0.66 from Rgb { r: 0, g: 0, b: 0 } to Rgb { r: 0, g: 0, b: 0 }
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/color.rs:35] Scaling RGB by 0.66 from Rgb { r: 213, g: 78, b: 83 } to Rgb { r: 140, g: 51, b: 54 }
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/color.rs:35] Scaling RGB by 0.66 from Rgb { r: 185, g: 202, b: 74 } to Rgb { r: 122, g: 133, b: 48 }
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/color.rs:35] Scaling RGB by 0.66 from Rgb { r: 230, g: 197, b: 71 } to Rgb { r: 151, g: 130, b: 46 }
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/color.rs:35] Scaling RGB by 0.66 from Rgb { r: 122, g: 166, b: 218 } to Rgb { r: 80, g: 109, b: 143 }
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/color.rs:35] Scaling RGB by 0.66 from Rgb { r: 195, g: 151, b: 216 } to Rgb { r: 128, g: 99, b: 142 }
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/color.rs:35] Scaling RGB by 0.66 from Rgb { r: 112, g: 192, b: 186 } to Rgb { r: 73, g: 126, b: 122 }
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/color.rs:35] Scaling RGB by 0.66 from Rgb { r: 234, g: 234, b: 234 } to Rgb { r: 154, g: 154, b: 154 }
[2019-12-16 14:45] [INFO] Initialisation complete
[2019-12-16 14:45] [INFO] Font size changed to Size(22) with DPR of 2
[2019-12-16 14:45] [DEBUG] Term::resize dimensions unchanged
[2019-12-16 14:45] [INFO] Width: 1600, Height: 1200
[2019-12-16 14:45] [DEBUG] [unhandled osc_dispatch]: [['1','3','3','7',],['R','e','m','o','t','e','H','o','s','t','=','b','r','i','c','3','@','_','_','_','E','D','I','T','E','D','_','_','_','_','_',],] at line 761
[2019-12-16 14:45] [DEBUG] [unhandled osc_dispatch]: [['1','3','3','7',],['C','u','r','r','e','n','t','D','i','r','=','/','U','s','e','r','s','/','b','r','i','c','3',],] at line 761
[2019-12-16 14:45] [DEBUG] [unhandled osc_dispatch]: [['1','3','3','7',],['S','h','e','l','l','I','n','t','e','g','r','a','t','i','o','n','V','e','r','s','i','o','n','=','6',],['s','h','e','l','l','=','z','s','h',],] at line 761
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1477] Carriage return
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1477] Carriage return
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1197] Setting window title to 'bric3@___EDITED_____: ~'
[2019-12-16 14:45] [DEBUG] [unhandled osc_dispatch]: [['1','3','3',],['D',],['0',],] at line 761
[2019-12-16 14:45] [DEBUG] [unhandled osc_dispatch]: [['1','3','3','7',],['R','e','m','o','t','e','H','o','s','t','=','b','r','i','c','3','@','_','_','_','E','D','I','T','E','D','_','_','_','_','_',],] at line 761
[2019-12-16 14:45] [DEBUG] [unhandled osc_dispatch]: [['1','3','3','7',],['C','u','r','r','e','n','t','D','i','r','=','/','U','s','e','r','s','/','b','r','i','c','3',],] at line 761
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1477] Carriage return
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Reset
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: CancelReverse
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: CancelUnderline
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1747] Clearing screen: Below
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Reset
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Background(Named(Background))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Named(Foreground))
[2019-12-16 14:45] [DEBUG] [unhandled osc_dispatch]: [['1','3','3',],['A',],] at line 761
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Reset
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Background(Named(Background))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Indexed(67))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Reset
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Indexed(67))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Background(Named(Background))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Indexed(67))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Reset
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Indexed(67))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Background(Named(Background))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Named(Black))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Reset
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Named(Black))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Background(Named(Background))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Named(Foreground))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Reset
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Background(Named(Background))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Named(Foreground))
[2019-12-16 14:45] [DEBUG] [unhandled osc_dispatch]: [['1','3','3',],['B',],] at line 761
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1660] Clearing line: Right
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1388] Moving forward: 100
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Reset
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Background(Named(Background))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Named(White))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Reset
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Named(White))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Background(Named(Background))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Named(White))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Reset
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Named(White))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Background(Named(Background))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Named(White))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1660] Clearing line: Right
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1839] Setting attribute: Foreground(Named(Foreground))
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1395] Moving backward: 117
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1870] Setting mode: CursorKeys
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1986] Setting keypad application mode
[2019-12-16 14:45] [TRACE] [alacritty_terminal/src/term/mod.rs:1870] Setting mode: BracketedPaste
```
</details>

**Comment by chrisduerr** [comment]
> I have the same issue using the Sauce Code Pro Nerd Font Complete Mono, the font seems to be picked, but not the special characters

That sounds like it might not actually be a fallback issue, since we will always pick from the primary font if the glyph is available.

Could you confirm that the specific powerline glyph is indeed present in that specific font @bric3?

It would be interesting to see what happens when the following patch is applied. That completely disables font fallback, so a missing glyph symbol should show up if this is a fallback issue.

```diff
diff --git a/font/src/ft/mod.rs b/font/src/ft/mod.rs
index 431a500..8a69975 100644
--- a/font/src/ft/mod.rs
+++ b/font/src/ft/mod.rs
@@ -306,22 +306,7 @@ impl FreeTypeRasterizer {
         glyph_key: GlyphKey,
         have_recursed: bool,
     ) -> Result<FontKey, Error> {
-        let c = glyph_key.c;
-
-        let use_initial_face = if let Some(face) = self.faces.get(&glyph_key.font_key) {
-            let index = face.ft_face.get_char_index(c as usize);
-
-            index != 0 || have_recursed
-        } else {
-            false
-        };
-
-        if use_initial_face {
-            Ok(glyph_key.font_key)
-        } else {
-            let key = self.load_face_with_glyph(c).unwrap_or(glyph_key.font_key);
-            Ok(key)
-        }
+        Ok(glyph_key.font_key)
     }
 
     fn get_rendered_glyph(&mut self, glyph_key: GlyphKey) -> Result<RasterizedGlyph, Error> {
```

**Comment by bric3** [comment]
For example I can see this glyph on iterm2. It's hex code is `F015`, could it be an UTF-8 issue since this character is encoded as 3 bytes in UTF-8 ?

```
  ~ echo  | hexdump
0000000 ef 80 95 0a
0000004
  ~ echo -e "\U0F015" | hexdump
0000000 ef 80 95 0a
0000004
```

<img width="152" alt="image" src="https://user-images.githubusercontent.com/803621/70916094-63df3600-201b-11ea-89e4-c1aac77a6a95.png">

The font is from https://github.com/ryanoasis/nerd-fonts

**Comment by kchibisov** [comment]
@bric3 what's your locale output? Could you launch alacritty from iterm2 and retest?

**Comment by bric3** [comment]
Here's my locale

```
  ~ locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL="en_US.UTF-8"
```

> Could you launch alacritty from iterm2 and retest?

I'm not sure what you mean?

**Comment by kchibisov** [comment]
@bric3 I was talking about launching alacritty executable from iterm2. You locale seems fine, however I'm not sure that I've ever seen a correct locale on macOS...

**Comment by bric3** [comment]
This has no effect, this is what I did to get the logs by the way `/Applications/Alacritty.app/Contents/MacOS/alacritty -vvv` in this https://github.com/jwilm/alacritty/issues/2657#issuecomment-566069080

**Comment by chrisduerr** [comment]
That patch is only useful on X11.

**Comment by kchibisov** [comment]
@bric3 Could you try changing your font name to this one `SauceCodePro Nerd Font Mono`?

**Comment by bric3** [comment]
Ah yes it works !

**Comment by kchibisov** [comment]
So alacritty was loading a wrong font for you, because you've entered a wrong font family name, and your issue is not related to the one we see on Linux/BSD systems.

FWIW we're not using fontconfig to load things on macOS, however it's a good idea to check your font family name using it, I guess.

**Comment by bric3** [comment]
@kchibisov Still thank you !!!
I think that alacritty could be improved in that way, i.e. to log which font it has loaded.

**Comment by chrisduerr** [comment]
@AkechiShiro, would you mind testing the patch I've linked earlier and showing the result of that?

I'm generally inclined to close this as duplicate of #1560, since I don't see any indication that this is something other than a font fallback issue.

**Comment by AkechiShiro** [comment]
@chrisduerr I just don't really know how can I apply the patch, I need to build Alacritty with this different code? How can I actually apply this patch ? 

> > I have the same issue using the Sauce Code Pro Nerd Font Complete Mono, the font seems to be picked, but not the special characters
> 
> That sounds like it might not actually be a fallback issue, since we will always pick from the primary font if the glyph is available.
> 
> Could you confirm that the specific powerline glyph is indeed present in that specific font @bric3?
> 
> It would be interesting to see what happens when the following patch is applied. That completely disables font fallback, so a missing glyph symbol should show up if this is a fallback issue.
> 
> ```diff
> diff --git a/font/src/ft/mod.rs b/font/src/ft/mod.rs
> index 431a500..8a69975 100644
> --- a/font/src/ft/mod.rs
> +++ b/font/src/ft/mod.rs
> @@ -306,22 +306,7 @@ impl FreeTypeRasterizer {
>          glyph_key: GlyphKey,
>          have_recursed: bool,
>      ) -> Result<FontKey, Error> {
> -        let c = glyph_key.c;
> -
> -        let use_initial_face = if let Some(face) = self.faces.get(&glyph_key.font_key) {
> -            let index = face.ft_face.get_char_index(c as usize);
> -
> -            index != 0 || have_recursed
> -        } else {
> -            false
> -        };
> -
> -        if use_initial_face {
> -            Ok(glyph_key.font_key)
> -        } else {
> -            let key = self.load_face_with_glyph(c).unwrap_or(glyph_key.font_key);
> -            Ok(key)
> -        }
> +        Ok(glyph_key.font_key)
>      }
>  
>      fn get_rendered_glyph(&mut self, glyph_key: GlyphKey) -> Result<RasterizedGlyph, Error> {
> ```

**Comment by chrisduerr** [comment]
Patches in git are applied using the `git apply <FILE>` command.

**Comment by madetara** [comment]
Hello, it seems I am having similar problems

OS: Windows 10 (18362) + WSL Ubuntu 18.04

The issue is reproduced in WSL shell

Upper part of the image is [Terminus](https://github.com/Eugeny/terminus)
Lower part of the image is Alacritty built with fallback patch applied
 
![изображение](https://user-images.githubusercontent.com/31938575/71983326-93a6e580-3237-11ea-9b94-800aa0c9917f.png)

Both emulators have:

- ConPTY enabled
- same font family
- TERM=cygwin

**Comment by kchibisov** [comment]
@madetara could you test #3127 ?

**Comment by madetara** [comment]
I forgot to mention that some symbols are displayed correctly 
![изображение](https://user-images.githubusercontent.com/31938575/71984139-53e0fd80-3239-11ea-9147-c2cdd234e7d0.png)

Script content:
```bash
#!/bin/bash

echo -e "Powerline glyphs:\n\
Code points Glyphe  Description                Old code point
U+E0A0      \xee\x82\xa0       Version control branch     (U+2B60 \xe2\xad\xa0 )\n\
U+E0A1      \xee\x82\xa1       LN (line) symbol           (U+2B61 \xe2\xad\xa1 )\n\
U+E0A2      \xee\x82\xa2       Closed padlock             (U+2B64 \xe2\xad\xa4 )\n\
U+E0B0      \xee\x82\xb0       Rightwards black arrowhead (U+2B80 \xe2\xae\x80 )\n\
U+E0B1      \xee\x82\xb1       Rightwards arrowhead       (U+2B81 \xe2\xae\x81 )\n\
U+E0B2      \xee\x82\xb2       Leftwards black arrowhead  (U+2B82 \xe2\xae\x82 )\n\
U+E0B3      \xee\x82\xb3       Leftwards arrowhead        (U+2B83 \xe2\xae\x83 )\n\
"
```

**Comment by kchibisov** [comment]
@madetara they could be in your primary font, as I pointed to test #3127 the font fallback isn't implemented on Windows at all in the current master, so you're limited to your primary font, however we'd be curious about your experience with the PR I've mentioned.

**Comment by madetara** [comment]
Branch from #3127 seems to be working!

Guess my only option is to use that branch?

**Comment by kchibisov** [comment]
@madetara yes, it's still requires changed to be merged upstream, as you can see, and there's no other option to make fontfallback to work on windows.

**Comment by shadowrylander** [comment]
Hello!
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Not to sound too pushy, however, may I inquire as to the status of this issue? Should I use the #3127 branch, and how would I go about installing it if I should, or should I wait for the master branch when the changes are merged?
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Thank you kindly for understanding!

**Comment by kchibisov** [comment]
The branch #3127 currently blocked by some upstream issues, so it wont build right now. So you should wait for a while until upstream progress. However I feel like you can just safely wait when it'll be in master.

**Comment by shadowrylander** [comment]
I'm assuming the `No ETA` rule applies here? :joy_cat:

**Comment by kchibisov** [comment]
it's opensource, there's no ETA :- ).

**Comment by shadowrylander** [comment]
Thought so! Thanks; I'll wait patiently for the master, then! :joy_cat:

**Comment by madetara** [comment]
I've built it 10 days ago when it was at 17946ec
Everything works fine

**Comment by kchibisov** [comment]
@madetara since then it was rebased/force pushed and as you can see it's not that easy to build it right now, due to https://github.com/alacritty/alacritty/pull/3127#issuecomment-574901592 , so it'll require you to prepare a proper environment, but it's possible to build ofc, just a bit of work, which can be hard if you're not familiar with rust.

**Comment by madetara** [comment]
If @shadowrylander is willing to try building I can push sources to my fork I think

**Comment by shadowrylander** [comment]
@kchibisov @madetara And I'm not familiar with Rust! :joy_cat: I'd have no idea where to begin!

**Comment by kchibisov** [comment]
Just to mention, the original issue was about Fontcofnig fallback, which is solved now, so all Windows folks should move to #3215

**Comment by shadowrylander** [comment]
Linux? :cat:

**Comment by Try2Code** [comment]
I had the same issue on Arch linux, but building it from the current repo solved the issue for me

**Comment by azarmadr** [comment]
i have the above issue
is there any proper config or solution i need

**Comment by ProgrammingLife** [comment]
Having the same issue with font: "DejaVu Sans Mono for Powerline". Powerline symbols don't show at all but this font works fine in urxvt in the same OS (Linux Mint).

**Comment by LucienBrule** [comment]
For what it's worth to anyone coming back to this, here's what I did.

1. Grab a patched font, in my case I wanted [JetBrains Mono](https://github.com/seanghay/JetBrainsMono-Powerline)
2. Download the release zip, extract to /usr/share/fonts/
3. Set the alacritty font stack to
```yaml
font:
  normal:
    family: JetBrains Mono
```

Despite the disclaimer that Jet Brains Mono supports powerline fonts, it did not on my setup thus the patched font was the way to go.
